File: packages/domain/lib/src/budget/budget_service.dart
import '../calendar/hijri_calendar.dart';
import '../category/category_tree.dart';
import '../ledger/ledger_entities.dart';
import '../money/currency_service.dart';

enum BudgetPeriod { gregorian, hijri }

class Budget {
  final String id;

  /// Target categories. Each counts itself plus its descendants.
  final List<String> categoryIds;

  /// Legacy single-category budgets carried a lone `categoryId`.
  final String? legacyCategoryId;

  final int amount; // minor units, in [currency]
  final String currency;
  final BudgetPeriod period;
  final bool rollover;

  const Budget({
    required this.id,
    this.categoryIds = const [],
    this.legacyCategoryId,
    required this.amount,
    required this.currency,
    this.period = BudgetPeriod.gregorian,
    this.rollover = false,
  });
}

class CategorySpend {
  final String categoryId;
  final String name;
  final String? color;
  final String? icon;
  final int spend;
  const CategorySpend({
    required this.categoryId,
    required this.name,
    this.color,
    this.icon,
    required this.spend,
  });
}

class EffectiveLimit {
  final int limit;
  final int rollover;
  const EffectiveLimit(this.limit, this.rollover);
}

typedef _Slice = ({String? categoryId, int amount, String currency});

/// Budget period-spend + rollover logic — a faithful Dart port of the JS
/// `BudgetService` (pure parts only; CRUD lives in the data layer).
///
/// `now` is injected per call (instead of `new Date()`), so the logic is pure
/// and deterministic in tests. Hijri periods read the transaction's immutable
/// `hijriDate` snapshot, falling back to a raw (offset-0) computation for legacy
/// rows — exactly as the JS did, so a later offset change never reclassifies a
/// past transaction.
class BudgetService {
  final CategoryTree categories;
  final CurrencyService fx;
  final HijriCalendar hijri;

  const BudgetService({
    required this.categories,
    required this.fx,
    required this.hijri,
  });

  List<String> targetCategoryIds(Budget b) {
    if (b.categoryIds.isNotEmpty) return b.categoryIds;
    final legacy = b.legacyCategoryId;
    return legacy != null ? [legacy] : const [];
  }

  Set<String> _expandedIds(Budget b) {
    final set = <String>{};
    for (final id in targetCategoryIds(b)) {
      set.addAll(categories.descendants(id));
    }
    return set;
  }

  List<_Slice> _categoryAmounts(LedgerTransaction t) {
    final splits = t.splits;
    if (splits != null && splits.isNotEmpty) {
      return [
        for (final s in splits)
          (categoryId: s.categoryId, amount: s.amount, currency: t.currency),
      ];
    }
    return [(categoryId: t.categoryId, amount: t.amount, currency: t.currency)];
  }

  bool _inCurrentPeriod(Budget b, LedgerTransaction t, DateTime now, int offset) {
    if (t.type != TxType.expense) return false;
    final d = t.date;
    if (d == null) return false;
    if (b.period == BudgetPeriod.hijri) {
      final todayH = hijri.toHijri(now, offset: offset);
      final h = t.hijriDate ?? hijri.toHijriRaw(d);
      return h.year == todayH.year && h.month == todayH.month;
    }
    return d.year == now.year && d.month == now.month;
  }

  int _sumMatching(
    Budget b,
    List<LedgerTransaction> txs,
    Set<String> ids,
    bool Function(LedgerTransaction) inPeriod,
  ) {
    var spend = 0;
    for (final t in txs) {
      if (!inPeriod(t)) continue;
      for (final slice in _categoryAmounts(t)) {
        final cid = slice.categoryId;
        if (cid != null && ids.contains(cid)) {
          spend += fx.convert(slice.amount, slice.currency, b.currency);
        }
      }
    }
    return spend;
  }

  /// Total current-period spend across all target categories (+ descendants).
  int currentSpend(Budget b, List<LedgerTransaction> txs, DateTime now,
      {int hijriOffset = 0}) {
    return _sumMatching(b, txs, _expandedIds(b),
        (t) => _inCurrentPeriod(b, t, now, hijriOffset));
  }

  /// Per-target-category spend for the current period.
  List<CategorySpend> spendByCategory(
      Budget b, List<LedgerTransaction> txs, DateTime now,
      {int hijriOffset = 0}) {
    return targetCategoryIds(b).map((cid) {
      final ids = categories.descendants(cid).toSet();
      final spend = _sumMatching(
          b, txs, ids, (t) => _inCurrentPeriod(b, t, now, hijriOffset));
      final cat = categories.find(cid);
      return CategorySpend(
        categoryId: cid,
        name: cat?.name ?? 'Category',
        color: cat?.color,
        icon: cat?.icon,
        spend: spend,
      );
    }).toList();
  }

  /// Current-period expense transactions counting toward the budget, newest
  /// first.
  List<LedgerTransaction> periodTransactions(
      Budget b, List<LedgerTransaction> txs, DateTime now,
      {int hijriOffset = 0}) {
    final ids = _expandedIds(b);
    final out = txs
        .where((t) =>
            _inCurrentPeriod(b, t, now, hijriOffset) &&
            _categoryAmounts(t)
                .any((s) => s.categoryId != null && ids.contains(s.categoryId)))
        .toList();
    final epoch = DateTime.fromMillisecondsSinceEpoch(0);
    out.sort((a, c) => (c.date ?? epoch).compareTo(a.date ?? epoch));
    return out;
  }

  /// Effective limit including optional rollover from the previous period.
  EffectiveLimit effectiveLimit(
      Budget b, List<LedgerTransaction> txs, DateTime now,
      {int hijriOffset = 0}) {
    if (!b.rollover) return EffectiveLimit(b.amount, 0);
    final ids = _expandedIds(b);

    bool prevMatches(LedgerTransaction t) {
      if (t.type != TxType.expense) return false;
      final d = t.date;
      if (d == null) return false;
      if (b.period == BudgetPeriod.hijri) {
        final todayH = hijri.toHijri(now, offset: hijriOffset);
        var pm = todayH.month - 1, py = todayH.year;
        if (pm < 0) {
          pm = 11;
          py -= 1;
        }
        final h = t.hijriDate ?? hijri.toHijriRaw(d);
        return h.year == py && h.month == pm;
      }
      final pm = now.month == 1 ? 12 : now.month - 1;
      final py = now.month == 1 ? now.year - 1 : now.year;
      return d.year == py && d.month == pm;
    }

    final prevSpent = _sumMatching(b, txs, ids, prevMatches);
    final leftover = (b.amount - prevSpent) > 0 ? b.amount - prevSpent : 0;
    return EffectiveLimit(b.amount + leftover, leftover);
  }
}

File: mobile-app/src/domain/services/BudgetService.js
/**
 * BudgetService — Budget CRUD, period spend calculation, rollover.
 *
 * No DOM, no side effects.  All state access via Store.
 */
import { Store }           from '../../core/Store.js';
import { IdGenerator }     from './IdGenerator.js';
import { CurrencyService } from './CurrencyService.js';
import { CategoryService } from './CategoryService.js';
import { HijriCalendarService } from './HijriCalendarService.js';

export class BudgetService {
  /** @type {Store} */               #store;
  /** @type {CurrencyService} */     #fx;
  /** @type {CategoryService} */     #cats;
  /** @type {HijriCalendarService} */ #hijri;

  constructor() {
    this.#store = Store.getInstance();
    this.#fx    = new CurrencyService();
    this.#cats  = new CategoryService();
    this.#hijri = new HijriCalendarService();
  }

  // ── Targeting ───────────────────────────────────────────────────────

  /**
   * The category IDs a budget targets. Backward-compatible: a legacy budget
   * carrying a single `categoryId` is treated as a one-element list.
   * @param {object} budget
   * @returns {string[]}
   */
  targetCategoryIds(budget) {
    if (Array.isArray(budget.categoryIds) && budget.categoryIds.length) return budget.categoryIds;
    return budget.categoryId ? [budget.categoryId] : [];
  }

  /** Every category that counts toward the budget = each target + its descendants. */
  #expandedIds(budget) {
    const set = new Set();
    for (const id of this.targetCategoryIds(budget)) {
      for (const d of this.#cats.descendants(id)) set.add(d);
    }
    return set;
  }

  /**
   * True if an expense tx falls within the budget's current period.
   *
   * For Hijri periods, t.hijriDate (the immutable snapshot written at creation
   * time) stays authoritative, so adjusting the offset never reclassifies past
   * transactions into a different budget month.
   *
   * The catch is that BOTH sides of the comparison must live in the same epoch.
   * "Today" used to be computed with the CURRENT offset while each snapshot
   * carried the offset in force when it was written, so nudging the offset near
   * a month boundary moved today into the next Hijri month while that same
   * day's transactions stayed in the previous one — and they silently dropped
   * out of the budget. #todayInEpochOf() evaluates today with the transaction's
   * own offset instead.
   */
  /**
   * The frozen Hijri date for a transaction.
   * Legacy rows with no snapshot pre-date the offset system, so offset 0.
   * @param {object} t
   * @returns {{year:number, month:number, day:number, offset:number}}
   */
  #hijriOf(t) {
    return t.hijriDate ?? this.#hijri.toHijriRaw(t.date);
  }

  /**
   * Today, evaluated in the SAME calendar epoch the transaction's snapshot was
   * written in — see the note on #inCurrentPeriod().
   *
   * A snapshot from before offsets were recorded has no `offset` field. Those
   * are treated as belonging to the user's current epoch, which reproduces the
   * previous behaviour for the overwhelmingly common case of an offset that
   * never changed, rather than silently re-bucketing old data.
   * @param {object} t
   * @returns {{year:number, month:number, day:number, offset:number}}
   */
  #todayInEpochOf(t) {
    const snap = t.hijriDate;
    const epoch = (snap && Number.isFinite(snap.offset))
      ? snap.offset
      : (snap ? this.#hijri.offset : 0);
    return this.#hijri.toHijriWithOffset(new Date(), epoch);
  }

  #inCurrentPeriod(budget, t) {
    if (t.type !== 'expense') return false;
    if (!t.date) return false; // guard against malformed transactions with missing date
    if (budget.period === 'hijri') {
      const h      = this.#hijriOf(t);
      const todayH = this.#todayInEpochOf(t);
      return h.year === todayH.year && h.month === todayH.month;
    }
    const now = new Date();
    const d   = new Date(t.date + 'T12:00:00');
    return d.getFullYear() === now.getFullYear() && d.getMonth() === now.getMonth();
  }

  // ── Queries ─────────────────────────────────────────────────────────

  /**
   * Total current-period spend for a budget across all its target categories
   * (and their descendants), in the budget's currency.
   * @param {object} budget
   * @returns {number} minor units
   */
  currentSpend(budget) {
    const state = this.#store.getState();
    const ids   = this.#expandedIds(budget);
    let spend = 0;
    for (const t of state.transactions) {
      if (!this.#inCurrentPeriod(budget, t)) continue;
      for (const slice of this.#categoryAmounts(t)) {
        if (ids.has(slice.categoryId)) spend += this.#fx.convert(slice.amount, slice.currency, budget.currency);
      }
    }
    return spend;
  }

  /**
   * Per-target-category spend for the current period (budget currency). Each
   * entry covers that target category plus its descendants — powering the small
   * split shown under a budget's progress bar.
   * @param {object} budget
   * @returns {{ categoryId: string, name: string, color: string, icon: string, spend: number }[]}
   */
  spendByCategory(budget) {
    const state = this.#store.getState();
    return this.targetCategoryIds(budget).map((cid) => {
      const ids = new Set(this.#cats.descendants(cid));
      let spend = 0;
      for (const t of state.transactions) {
        if (!this.#inCurrentPeriod(budget, t)) continue;
        for (const slice of this.#categoryAmounts(t)) {
          if (ids.has(slice.categoryId)) spend += this.#fx.convert(slice.amount, slice.currency, budget.currency);
        }
      }
      const cat = this.#cats.find(cid);
      return { categoryId: cid, name: cat?.name || 'Category', color: cat?.color || '#a1a1aa', icon: cat?.icon || 'circle', spend };
    });
  }

  /**
   * All expense transactions counting toward the budget this period, newest
   * first — used by the budget detail view.
   * @param {object} budget
   * @returns {object[]}
   */
  periodTransactions(budget) {
    const state = this.#store.getState();
    const ids   = this.#expandedIds(budget);
    return state.transactions
      .filter((t) => this.#inCurrentPeriod(budget, t) &&
                     this.#categoryAmounts(t).some((s) => ids.has(s.categoryId)))
      .sort((a, b) => b.date.localeCompare(a.date));
  }

  /**
   * Effective budget limit accounting for optional rollover from previous period.
   * @param {object} budget
   * @returns {{ limit: number, rollover: number }}
   */
  effectiveLimit(budget) {
    if (!budget.rollover) return { limit: budget.amount, rollover: 0 };

    const state     = this.#store.getState();
    const ids       = this.#expandedIds(budget);
    const now       = new Date();

    const prevMatches = (t) => {
      if (t.type !== 'expense') return false;
      if (!t.date) return false; // guard against missing date
      if (budget.period === 'hijri') {
        // Evaluate "last month" in this transaction's own epoch, for the same
        // reason #inCurrentPeriod() does — otherwise the rollover window drifts
        // away from the transactions it is meant to cover.
        const ref = this.#todayInEpochOf(t);
        let pm = ref.month - 1, py = ref.year;
        if (pm < 0) { pm = 11; py -= 1; }
        const h = this.#hijriOf(t);
        return h.year === py && h.month === pm;
      }
      const d  = new Date(t.date + 'T12:00:00');
      const pm = now.getMonth() === 0 ? 11 : now.getMonth() - 1;
      const py = now.getMonth() === 0 ? now.getFullYear() - 1 : now.getFullYear();
      return d.getFullYear() === py && d.getMonth() === pm;
    };

    let prevSpent = 0;
    for (const t of state.transactions) {
      if (!prevMatches(t)) continue;
      for (const slice of this.#categoryAmounts(t)) {
        if (ids.has(slice.categoryId)) {
          prevSpent += this.#fx.convert(slice.amount, slice.currency, budget.currency);
        }
      }
    }

    const leftover = Math.max(0, budget.amount - prevSpent);
    return { limit: budget.amount + leftover, rollover: leftover };
  }

  // ── Private helpers ──────────────────────────────────────────────────

  #categoryAmounts(tx) {
    if (Array.isArray(tx.splits) && tx.splits.length) {
      return tx.splits.map((s) => ({
        categoryId: s.categoryId || null,
        amount:     s.amount,
        currency:   tx.currency,
      }));
    }
    return [{ categoryId: tx.categoryId || null, amount: tx.amount, currency: tx.currency }];
  }

  // ── CRUD ─────────────────────────────────────────────────────────────

  /**
   * Create a budget.
   * @param {object} data
   * @returns {object}
   */
  create(data) {
    const ids = Array.isArray(data.categoryIds) && data.categoryIds.length
      ? data.categoryIds
      : (data.categoryId ? [data.categoryId] : []);
    const budget = {
      id:          IdGenerator.generate('bg'),
      categoryIds: ids,
      categoryId:  ids[0] ?? null,   // kept in sync for backward compatibility
      amount:      data.amount,
      currency:    data.currency,
      period:      data.period === 'hijri' ? 'hijri' : 'gregorian',
      rollover:    data.rollover || false,
    };
    this.#store.getState().budgets.push(budget);
    this.#store.flush();
    return budget;
  }

  /**
   * Update a budget.
   * @param {string} id
   * @param {object} changes
   */
  update(id, changes) {
    const budget = this.#store.getState().budgets.find((b) => b.id === id);
    if (!budget) return null;
    Object.assign(budget, changes);
    this.#store.flush();
    return budget;
  }

  /**
   * Delete a budget.
   * @param {string} id
   */
  delete(id) {
    const state = this.#store.getState();
    state.budgets = state.budgets.filter((b) => b.id !== id);
    this.#store.flush();
  }
}
